File: Interv02/06.py
# https://leetcode.cn/problems/palindrome-linked-list-lcci/

# Definition for singly-linked list.

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class LinkList:
    def __init__(self):
        self.__head = None
        self.__length = 0

    def __repr__(self):
        cur = self.__head
        lis = []
        for i in range(self.__length):
            lis.append(str(cur.val))
            cur = cur.next
        return " -> ".join(lis)

    def append(self, item):
        node = ListNode(item)
        if self.__length == 0:
            self.__head = node
        else:
            cur = self.__head
            while cur.next:
                cur = cur.next
            cur.next = node
        self.__length += 1

    def get_head(self):
        return self.__head

    def update_length(self):
        self.__length = 0
        cur = self.__head
        while cur:
            self.__length += 1
            cur = cur.next


class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        cur = head
        length = 0
        while cur:
            length += 1
            cur = cur.next
        fast, low = head, head
        length = length // 2 if length % 2 == 0 else length // 2 + 1
        for i in range(length):
            fast = fast.next
        before, after = [], []
        while fast:
            before.append(low.val)
            after.append(fast.val)
            fast, low = fast.next, low.next
        return before == after[::-1]


if __name__ == '__main__':
    l = LinkList()
    arr = [1, 4, -1, -1, 4, 1]
    for item in arr:
        l.append(item)
    print(l)
    test = Solution()
    print(test.isPalindrome(l.get_head()))

File: Interv02/06.cpp
//
// Created by Air on 2023/2/28.
//
// https://leetcode.cn/problems/palindrome-linked-list-lcci/

#include "LinkList.h"
#include "stack"
#include "iostream"
#include "vector"

class Solution {
public:
    // 采用栈解决
    bool isPalindrome(ListNode* head) {
        int length(0);
        ListNode* cur(head);
        while (cur) {
            ++length;
            cur = cur->next;
        }

        stack<int> s;
        cur = head;
        for (int i = 0; i < length / 2; ++i) {
            s.push(cur->val);
            cur = cur->next;
        }
        if (length % 2)
            cur = cur->next;
        bool flag(true);
        for (int i = 0; i < length / 2; ++i) {
            if (s.top() != cur->val) {
                flag = false;
                break;
            }
            s.pop();
            cur = cur->next;
        }
        return flag;
    }
};


int main() {
    LinkList L;
    vector<int> arr = {1, 4, 3, 3, 4, 1};
    for (int item : arr)
        L.push_back(item);
    cout << L.to_string() << endl;

    Solution test;
    cout << test.isPalindrome(L.get_head());
    return 0;
}